pkg/build/generator/generator.go

```diff
@@ -0,0 +1,243 @@
+	buildapi "github.com/openshift/origin/pkg/build/api"
+	imageapi "github.com/openshift/origin/pkg/image/api"
+)
+
+// BuildGenerator is a central place responsible for generating new Build objects
+// from BuildConfigs and other Builds.
+type BuildGenerator struct {
+	Client GeneratorClient
+}
+
+type GeneratorClient interface {
+	GetBuildConfig(ctx kapi.Context, name string) (*buildapi.BuildConfig, error)
+	UpdateBuildConfig(ctx kapi.Context, buildConfig *buildapi.BuildConfig) error
+	GetBuild(ctx kapi.Context, name string) (*buildapi.Build, error)
+	CreateBuild(ctx kapi.Context, build *buildapi.Build) error
+	GetImageRepository(ctx kapi.Context, name string) (*imageapi.ImageRepository, error)
+}
+
+type Client struct {
+	GetBuildConfigFunc     func(ctx kapi.Context, name string) (*buildapi.BuildConfig, error)
+	UpdateBuildConfigFunc  func(ctx kapi.Context, buildConfig *buildapi.BuildConfig) error
+	GetBuildFunc           func(ctx kapi.Context, name string) (*buildapi.Build, error)
+	CreateBuildFunc        func(ctx kapi.Context, build *buildapi.Build) error
+	GetImageRepositoryFunc func(ctx kapi.Context, name string) (*imageapi.ImageRepository, error)
+}
+
+func (c Client) GetBuildConfig(ctx kapi.Context, name string) (*buildapi.BuildConfig, error) {
+	return c.GetBuildConfigFunc(ctx, name)
+}
+
+func (c Client) UpdateBuildConfig(ctx kapi.Context, buildConfig *buildapi.BuildConfig) error {
+	return c.UpdateBuildConfigFunc(ctx, buildConfig)
+}
+
+func (c Client) GetBuild(ctx kapi.Context, name string) (*buildapi.Build, error) {
+	return c.GetBuildFunc(ctx, name)
+}
+
+func (c Client) CreateBuild(ctx kapi.Context, build *buildapi.Build) error {
+	return c.CreateBuildFunc(ctx, build)
+}
+
+func (c Client) GetImageRepository(ctx kapi.Context, name string) (*imageapi.ImageRepository, error) {
+	return c.GetImageRepositoryFunc(ctx, name)
+}
+
+// Generate returns new Build object based on BuildRequest object
+func (g *BuildGenerator) Generate(ctx kapi.Context, request *buildapi.BuildRequest) (*buildapi.Build, error) {
+	var newBuild *buildapi.Build
+	if len(request.BuildConfigName) > 0 {
+		glog.Infof("Generating build for config %s", request.BuildConfigName)
+		bc, err := g.Client.GetBuildConfig(ctx, request.BuildConfigName)
+		if err != nil {
+			return nil, err
+		}
+		newBuild, err = g.generateBuildFromConfig(ctx, bc, request.Revision)
+		if err != nil {
+			return nil, err
+		}
+		err = g.Client.UpdateBuildConfig(ctx, bc)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		glog.Infof("Generating build from build %s", request.BuildName)
+		build, err := g.Client.GetBuild(ctx, request.BuildName)
+		if err != nil {
+			return nil, err
+		}
+		newBuild = generateBuildFromBuild(build)
+	}
+
+	if !kapi.ValidNamespace(ctx, &newBuild.ObjectMeta) {
+		return nil, errors.NewConflict("build", newBuild.Namespace, fmt.Errorf("Build.Namespace does not match the provided context"))
+	}
+	kapi.FillObjectMetaSystemFields(ctx, &newBuild.ObjectMeta)
```

Comment:
The build created here still is vulnerable to the name length problem (<63) should I add a logic to prevent that or at least validate that?